File: Day002_2025/Q002_GFG.py
class Solution:
    def countSubarrays(self, arr, k):
        # Q. Subarrays with sum K
        ans = 0
        curr_sum = 0
        prefix_sum = {0: 1}

        for num in arr:
            curr_sum += num
            if curr_sum - k in prefix_sum:
                ans += prefix_sum[curr_sum - k]
            prefix_sum[curr_sum] = prefix_sum.get(curr_sum, 0) + 1
        
        return ans


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        k = int(input())
        arr = list(map(int, input().split()))
        
        sol = Solution()
        print(sol.countSubarrays(arr, k))
        print("~")

File: Day002_2025/Q002_GFG.java
import java.util.*;

class Q002_GFG {
    public int countSubarrays(int[] arr, int k) {
        // Q. Subarrays with sum K
        int ans = 0, currSum = 0;
        HashMap<Integer, Integer> um = new HashMap<>();
        um.put(0, 1);

        for (int num : arr) {
            currSum += num;
            if (um.containsKey(currSum - k)) {
                ans += um.get(currSum - k);
            }
            um.put(currSum, um.getOrDefault(currSum, 0) + 1);
        }
        return ans;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = Integer.parseInt(sc.nextLine());
        while (t-- > 0) {
            int k = Integer.parseInt(sc.nextLine());
            String[] input = sc.nextLine().split(" ");
            int[] arr = Arrays.stream(input).mapToInt(Integer::parseInt).toArray();

            Q002_GFG sol = new Q002_GFG();
            System.out.println(sol.countSubarrays(arr, k));
            System.out.println("~");
        }
        sc.close();
    }
}
